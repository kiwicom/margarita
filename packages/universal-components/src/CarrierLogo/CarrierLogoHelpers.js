// @flow

import { SIZE_OPTIONS, CARRIER_TYPE_OPTIONS, BASE_URL } from './consts';
import type { CarrierData } from './CarrierLogoTypes';

export const getCarrierImageUri = (
  carriersLength: number,
  size: string,
  pixelRatio: number,
  carrierData: CarrierData,
): string => {
  let urlSize = carriersLength > 1 || size === SIZE_OPTIONS.SMALL ? 16 : 32;
  if (pixelRatio > 2.25) {
    urlSize *= 4;
  } else if (pixelRatio > 1.25) {
    urlSize *= 2;
  }

  return `${BASE_URL}/airlines/${urlSize}/${carrierData.code}.png?default=${
    carrierData.type === undefined
      ? CARRIER_TYPE_OPTIONS.AIRLINE
      : carrierData.type
  }.png`;
};

export const parseCarriers = (carriers: CarrierData[]): CarrierData[] => {
  const codeMap = carriers.map(carrierData => carrierData.code);
  return carriers
    .filter((carrierData, index) => codeMap.indexOf(carrierData.code) === index)
    .slice(0, 4);
};
