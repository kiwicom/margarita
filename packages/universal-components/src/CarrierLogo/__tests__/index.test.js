// @flow

import * as React from 'react';
import { Image } from 'react-native';
import { render } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';

import { CarrierLogo } from '..';

import { getCarrierImageUri, parseCarriers } from '../CarrierLogoHelpers';
import { SIZE_OPTIONS } from '../consts';

describe('CarrierLogo', () => {
  const carriers = [
    { code: 'FR', name: 'Ryanair' },
    { code: 'TO', name: 'Transavia France' },
    { code: 'VY', name: 'Vueling' },
    { code: 'OK', name: 'Czech Airlines' },
  ];

  it('should contain correct number of images', () => {
    const { getAllByType } = render(<CarrierLogo carriers={carriers} />);
    expect(getAllByType(Image)).toHaveLength(4);
  });

  it('should contain correct number of images with correct size when duplicate carriers data are supplied', () => {
    const { getAllByType, getByType } = render(
      <CarrierLogo carriers={[carriers[0], carriers[0]]} />,
    );
    expect(getAllByType(Image)).toHaveLength(1);
    expect(getByType(Image).props.source.uri).toMatch(/64/);
  });

  it('should match snapshot diff', () => {
    const base = <CarrierLogo carriers={[carriers[0]]} />;
    const extend = <CarrierLogo carriers={carriers} />;
    expect(snapshotDiff(base, extend)).toMatchSnapshot();
  });

  test('getCarrierImageUri > should return correct path', () => {
    const imageUri = getCarrierImageUri(1, SIZE_OPTIONS.LARGE, 3, carriers[1]);
    expect(imageUri).toMatchSnapshot();
  });

  test('getCarrierImageUri > should have high resolution', () => {
    expect(getCarrierImageUri(1, SIZE_OPTIONS.LARGE, 3, carriers[0])).toMatch(
      /128/,
    );
  });

  test('getCarrierImageUri > should return small size version in case of multiple carriers', () => {
    expect(getCarrierImageUri(4, SIZE_OPTIONS.LARGE, 1, carriers[0])).toMatch(
      /16/,
    );
  });

  test('parseCarriers > should return correct number of carriers', () => {
    expect(parseCarriers(carriers)).toHaveLength(4);
  });

  test('parseCarriers > should remove duplicate carriers', () => {
    const duplicateCarriers = [
      { code: 'FR', name: 'Ryanair' },
      { code: 'FR', name: 'Ryanair' },
      { code: 'TO', name: 'Transavia France' },
      { code: 'TO', name: 'Transavia France' },
    ];
    expect(parseCarriers(duplicateCarriers)).toHaveLength(2);
  });

  test('parseCarriers > should return max 4 carriers', () => {
    const longCarriers = [
      { code: 'FR', name: 'Ryanair' },
      { code: 'TO', name: 'Transavia France' },
      { code: 'VY', name: 'Vueling' },
      { code: 'OK', name: 'Czech Airlines' },
      { code: 'DI', name: 'Something' },
    ];
    expect(parseCarriers(longCarriers)).toHaveLength(4);
  });
});
