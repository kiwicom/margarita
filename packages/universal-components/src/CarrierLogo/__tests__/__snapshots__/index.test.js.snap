// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarrierLogo getCarrierImageUri > should return correct path 1`] = `"https://images.kiwi.com/airlines/128/TO.png?default=airline.png"`;

exports[`CarrierLogo should match snapshot diff 1`] = `
"Snapshot Diff:
- <CarrierLogo carriers={[{\\"code\\": \\"FR\\", \\"name\\": \\"Ryanair\\"}]} />
+ <CarrierLogo carriers={[{\\"code\\": \\"FR\\", \\"name\\": \\"Ryanair\\"}, {\\"code\\": \\"TO\\", \\"name\\": \\"Transavia France\\"}, {\\"code\\": \\"VY\\", \\"name\\": \\"Vueling\\"}, {\\"code\\": \\"OK\\", \\"name\\": \\"Czech Airlines\\"}]} />

@@ -10,29 +10,91 @@
        },
        Object {
          \\"height\\": 32,
          \\"width\\": 32,
        },
+       Object {
+         \\"alignContent\\": \\"space-between\\",
+       },
+     ]
+   }
+ >
+   <Image
+     source={
+       Object {
+         \\"uri\\": \\"https://images.kiwi.com/airlines/32/FR.png?default=airline.png\\",
+       }
+     }
+     style={
+       Array [
+         Object {
+           \\"backgroundColor\\": \\"transparent\\",
+           \\"borderRadius\\": 3,
+         },
+         Object {
+           \\"height\\": 15,
+           \\"width\\": 15,
+         },
          false,
        ]
      }
- >
+   />
    <Image
      source={
        Object {
-         \\"uri\\": \\"https://images.kiwi.com/airlines/64/FR.png?default=airline.png\\",
+         \\"uri\\": \\"https://images.kiwi.com/airlines/32/TO.png?default=airline.png\\",
        }
      }
      style={
        Array [
          Object {
            \\"backgroundColor\\": \\"transparent\\",
            \\"borderRadius\\": 3,
          },
          Object {
-           \\"height\\": 32,
-           \\"width\\": 32,
+           \\"height\\": 15,
+           \\"width\\": 15,
+         },
+         false,
+       ]
+     }
+   />
+   <Image
+     source={
+       Object {
+         \\"uri\\": \\"https://images.kiwi.com/airlines/32/VY.png?default=airline.png\\",
+       }
+     }
+     style={
+       Array [
+         Object {
+           \\"backgroundColor\\": \\"transparent\\",
+           \\"borderRadius\\": 3,
+         },
+         Object {
+           \\"height\\": 15,
+           \\"width\\": 15,
+         },
+         false,
+       ]
+     }
+   />
+   <Image
+     source={
+       Object {
+         \\"uri\\": \\"https://images.kiwi.com/airlines/32/OK.png?default=airline.png\\",
+       }
+     }
+     style={
+       Array [
+         Object {
+           \\"backgroundColor\\": \\"transparent\\",
+           \\"borderRadius\\": 3,
+         },
+         Object {
+           \\"height\\": 15,
+           \\"width\\": 15,
          },
          false,
        ]
      }
    />"
`;
